### DISCUSSION TITLE
Clustering issues when using data imported from ArchR

### DISCUSSION THREAD

**Discussion by laurajohanne** [aertslab/scenicplus#266] [Q&A] [answered]
Hi everyone, 

I have been using the tutorial https://scenicplus.readthedocs.io/en/latest/pbmc_multiome_tutorial.html#Tutorial:-10x-multiome-pbmc for my own dataset.

I am having some issues with the clustering step (UMAP) when using data from an ArchR project, where the cluster do not cluster based on cell annotation.

I am working with an ArchR project with 64654 cells.

> # Obtaining count matrix (for peaks) by converting the count matrix to a h5ad file formate (due to size) ;
>  matrix  <- ArchRtoSignac::getPeakMatrix(ArchRProject= ArchRPro)
>  scenic_plus_obj <- Seurat::CreateSeuratObject(counts = matrix, project = "count_matrix")
>  sceasy::convertFormat(scenic_plus_obj, from="seurat", to ="anndata",
                       outFile='/work/ArchR/data_trine_for_scenicplus/scenic_plus_obj.h5ad')

Next I follow the tutorial; https://pycistopic.readthedocs.io/en/latest/Toy_melanoma-RTD.html with the exception of when creating the cisTopic object since the count matrix is a Sparse matrix.

> # Create cisTopic object
> cistopic_obj = create_cistopic_object(fragment_matrix= count_matrix, cell_names = cell_annotations_list , region_names = region_annotations_list, path_to_blacklist='/work/Scenicplus_pipeline/data/mm10-blacklist.v2.bed')

> # Adding cell information
> cistopic_obj.add_cell_data(cell_data)

> pickle.dump(cistopic_obj,
            open(os.path.join(outDir, 'cistopic_obj.pkl'), 'wb'))
> print(cistopic_obj)

># Run models
>models=run_cgs_models_mallet(path_to_mallet_binary,
                    cistopic_obj,
                    n_topics=[2,5,10,15,20,25,30,35,40,45,50],
                    n_cpu=55,
                    n_iter=500,
                    random_state=1,
                    alpha=50,
                    alpha_by_topic=True,
                    eta=0.1,
                    eta_by_topic=False,
                    tmp_path= tmpDir, #Use SCRATCH if many models or big data set
                    save_path = None)

>with open(outDir+'Mallet_models_500.pkl', 'wb') as f:
>pickle.dump(models, f)

> model=evaluate_models(models,
                     select_model=None,
                     return_model=True,
                     metrics=['Arun_2010','Cao_Juan_2009', 'Minmo_2011', 'loglikelihood'],
                     plot_metrics= False,
                     save= outDir + 'models/model_selection.pdf')

> cistopic_obj.add_LDA_model(model)

<img width="801" alt="Skærmbillede 2023-12-06 kl  14 01 53" src="https://github.com/aertslab/scenicplus/assets/125036200/313acc24-092d-43f5-a77d-7678f1192118">

>  # Find clusters
> from pycisTopic.clust_vis import *
> find_clusters(cistopic_obj,
                 target  = 'cell',
                 k = 40,
                 res = [0.6, 1.2],
                 seed = 1,
                 prefix = 'pycisTopic_',
                 scale = True,
                 split_pattern = '-')

> # UMAP
>from pycisTopic.clust_vis import *
>run_umap(cistopic_obj, target  = 'cell', scale = True, random_state = 1)

>plot_metadata(cistopic_obj, 
              reduction_name = 'UMAP', 
              variables = ['OrderCelltype_merged', 'pycisTopic_leiden_40_0.6', 'pycisTopic_leiden_40_1.2'],
              num_columns=3,
                 text_size=10,
                 dot_size=0.5,
                 figsize=(15,7),
                 save= '/work/Scenicplus_pipeline/dataoutput/visualization/dimensionality_reduction_label_uncorrected_merged_merged.pdf')

I do not encounter any issues in the next steps of the tutorial until I reach visualization with UMAP where I am having issues.

When using the cell annotation from the ArchR project the cells seems to be randomly clustered together. When using their own clustering method leiden they cluster nicely visualize but it does not match the cell annotation.

<img width="1135" alt="Skærmbillede 2023-12-06 kl  13 44 37" src="https://github.com/aertslab/scenicplus/assets/125036200/0465034e-f076-4d64-b76b-7a8b143aedb6">

When using ArchR to cluster the data it also appear that the annotation is correct. The data have already been preprocessed and used by multiple people and I am confident that the annotation is correct.

> # UMAP from ArchR
<img width="314" alt="Skærmbillede 2023-12-06 kl  13 46 49" src="https://github.com/aertslab/scenicplus/assets/125036200/1e6c2692-5b30-4304-9878-f292b35c2577">

And when using the UMAP coordinates from ArchR within pycisTopic it appear that the two methods find quite different coordinates.

> # UMAP with coordinates from ArchR in pycisTopic
<img width="998" alt="Skærmbillede 2023-12-06 kl  13 49 28" src="https://github.com/aertslab/scenicplus/assets/125036200/a25b9966-c54f-44b9-9ab5-abb857a00966">

I have already checked if the order of the cells are the same in both the ArchR project and the cisTopic object, which they are. And I have used same seed parameter and number of neighbors within both methods. 

Does anyone might have an idea what my next step should be?

Thanks in advance,

Laura

**Comment by SeppeDeWinter** [comment]
Hi Laura

This is very strange. It really looks like the cells are not labeled correctly (i.e. their order is switched), even though you made sure this is not the case.

The reason I'm saying this is that the pycisTopic based UMAP and the ArchR based one in fact do look similar (i.e. the structure is more or less the same).

Some things you can check:

This will show the UMAP coordinates along with the cell barcodes, can you compare this to ArchR?

```python

cistopic_obj.projections["cell"]["UMAP"]

```

This will show the metadata that is present in the cisTopic obj, can you make sure that each barcode (index of this dataframe) is annotated the same in cisTopic and Arch.

```python

cistopic_obj.cell_data

```

You can also make sure wether there are any barcode collisions, i.e. are there duplicate barcodes in your cisTopic object? This can happen when the sample id was not appended correctly to each barcode.

I hope these tips can help you figure out what went wrong.

Feel free to ask questions.

All the best,
Seppe

**Comment by laurajohanne** [comment, marked as answer]
Hi Seppe,

Thank you for your reply and your suggestions.

UMAP coordinates along with the cell barcodes from cisTopic;
![Skærmbillede 2023-12-18 kl  10 35 05](https://github.com/aertslab/scenicplus/assets/125036200/0ab16bd2-1aa2-4b4f-be8f-e7e504d8015a)


UMAP coordinates along with the cell barcodes from ArchR;
![Skærmbillede 2023-12-18 kl  10 33 56](https://github.com/aertslab/scenicplus/assets/125036200/30ff51a0-49ba-43df-9094-afa11f116806)


cistopic_obj.cell_data;
![Skærmbillede 2023-12-18 kl  10 36 21](https://github.com/aertslab/scenicplus/assets/125036200/074b85a8-28e1-4aae-8e8d-f8aa38f8536c)

I have also checked if there are duplicate barcodes in the cisTopic object;
![Skærmbillede 2023-12-18 kl  10 43 15](https://github.com/aertslab/scenicplus/assets/125036200/f90f9daf-f810-4e06-bdb1-5ea8f1da4769)


And in the metadata;
![Skærmbillede 2023-12-18 kl  10 40 58](https://github.com/aertslab/scenicplus/assets/125036200/10bfe491-a615-42e3-b983-0b267f3e108d)

I am theorizing that something goes wrong when getting the UMAP coordinates in cistopic. I am now trying to another way to extract the data from ArchR (with Signac) and using generating a new  gene activity matrix using Signac functions to the SeuratObject instead of transferring the ArchR gene score matrix to the SeuratObject.

Best regards;
Laura
